Add ephemeral  dynamic runner workflow

pavlovic-ivan opened the pull request:
**Intention**

The intention of this PR is to provide the project the ability to deploy dynamic runners that will run CUDA jobs, by adding new runners workflow. In order to have the complete feature, Github repo needs to be configured. More about that bellow in the Guide.

**Guide**

This guide explains how to configure Github repo. First, the prerequisits. You need a protected environment. If none exist:

1. Go to `Settings` > `Environments` > `New environment`
2. Give the Environment a name `cuda-test-infra`
3. Select a branch for which this Environment will be available by selecting the option "Selected branches" in the drop down menu of `Deployment branches`.
4. Select "+ Add deployment branch rule"
5. In the popup window enter the name of the main branch (eg, main, master, whichever is the main for your repo)

Now, make sure you have a Personal Access Token (PAT) set. If not:

1. Open Your Profile > `Settings` > `Developer Settings` > `Personal access tokens`
2. Select `Generate new token`
3. Give the token a name
4. Set expiration to "No expiration"
5. Select scopes for the token: `repo: Full control of private repositories`
6. Select `Generate token`
7. Copy the token and save it, it is visible to you only once

After the Environment [exists/is created], add new Environment secrets (provided safely offline), but:

- Select "+ Add Secret" within selected Environment (Environment named `cuda-test-infra`)

**Working example**

A working example can be seen [here](https://github.com/pavlovic-ivan/ILGPU-fork2/actions/runs/1848580697). You will see that CUDA tests are green. Workflow failed on `publish-preview` job, as i don't have perms and secrets needed to make it successful.

When this workflow was created, it created an event that triggered a workflow that deploys the runners. Workflow available [here](https://github.com/pavlovic-ivan/ILGPU-fork2/actions/runs/1848580754). Once the main CI workflow is done, another event triggers a workflow that destroys the created runners (as seen [here](https://github.com/pavlovic-ivan/ILGPU-fork2/actions/runs/1848719253)).

**Further steps before accepting PR**

- [ ] Communicate securely with upstream maintainer. Provide secrets needed to deploy runners to GR GCP project

ljubon commented:
Nicely done Ivane! Learned a few neat stuff from your PR.. thank you! 👍 🥇

jgiannuzzi commented on .github/workflows/ci.yaml:
```diff
@@ -102,7 +118,7 @@ jobs:
     # Do not run this job for pull requests where both branches are from the same repo.
     # Jobs that depend on this one will be skipped too.
-    # This prevents duplicate CI runs for our own pull requests, whilst preserving the ability to
+    # This prevents duplicate CI runs for our own pull requests, whilst preserving the ability to 
```
```suggestion
    # This prevents duplicate CI runs for our own pull requests, whilst preserving the ability to
```

pavlovic-ivan replied:
Should be resolved now

jgiannuzzi commented on .github/workflows/ci.yaml:
```diff
@@ -159,8 +175,8 @@ jobs:
 
-      - name: Set values
-        id: values
+      - name: Prepare environment
+        id: prepare-env
         shell: bash
         run: |
           echo "::set-output name=json::{\"ILGPU_CLEAN_TESTS\":\"false\", \"ILGPU_TEST_PARALLELIZATION_LEVEL\":\"1\"}"
```
This should not be part of this PR, as it should be covered by https://github.com/m4rs-mt/ILGPU/pull/712
@ljubon could you please update the PR so it can be reviewed and merged? (add description, change job name and id, set empty json as default, echo only once)

pavlovic-ivan replied:
All right. Then, the #712 should be merged first, before this one. I will need to wait for that, no? I see it is currently a draft. @ljubon please add this PR to Ora board as well if it's not there

ljubon replied:
PR updated https://github.com/m4rs-mt/ILGPU/pull/712 🤞

jgiannuzzi replied:
@pavlovic-ivan I see that the `master` branch of your fork is not in line with the upstream repo, but rather has an old version of @ljubon's branch for https://github.com/m4rs-mt/ILGPU/pull/712.

Could you please either update your `master` branch, or reset it to the latest upstream `master`?

jgiannuzzi commented on .github/workflows/ci.yaml:
```diff
@@ -433,4 +449,4 @@ jobs:
 
       - name: Publish to NuGet
-        run: dotnet nuget push "*.nupkg" --api-key ${{ secrets.NUGET_API_KEY }} --source https://api.nuget.org/v3/index.json
+        run: dotnet nuget push "*.nupkg" --api-key ${{ secrets.NUGET_API_KEY }} --source https://api.nuget.org/v3/index.json
```
Make sure you don't change the existing formatting

pavlovic-ivan replied:
Not sure if new line is the problem or not. I actually didn't change here anything. Could be that different new line characters were introduced automatically by my OS (if it is different than Marcel's)

jgiannuzzi replied:
It's your editor that is acting differently from Marcel's. It's best to avoid those changes, as they detract from the actual content of the pull request. It would also create a flip-flop of formatting between contributors.
One way to solve this is to rewrite your commit with [GitUp](https://gitup.co) and discard this line change.

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,123 @@
+          
+          (
+            [ "${{ github.event.repository.fork }}" == "false" ] || 
+            [ $GITHUB_REPOSITORY == "${{ github.repository }}" ]
```
I _think_ this will always evaluate to `true`. Is there a reason why you don't use the same logic as in `ci.yml` to check whether it's a fork?

jgiannuzzi replied:
Also worth checking whether this is needed at all — maybe `github.event.repository.fork` is enough for `workflow_run` events?

pavlovic-ivan replied:
I'll check, thanks

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,123 @@
+    if: ${{ github.event.action == 'requested' && needs.is-fork.outputs.upstream }}
+    environment: cuda-test-infra
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Prepare artifacts
+        run: |
+          mkdir -p artifacts
+          cp config.yaml artifacts
+
+      - name: Upload config
+        uses: actions/upload-artifact@v2
+        with:
+          name: pulumi-config
+          path: artifacts
+          retention-days: 1
+
+      - name: Checkout Pulumi Ephemeral Github Runner project
+        uses: actions/checkout@v2
+        with:
+          repository: pavlovic-ivan/ephemeral-github-runner-gcp
+          ref: main
+
+      - name: Download config
+        uses: actions/download-artifact@v2
+        with:
+          name: pulumi-config
+          path: artifacts
```
Instead of uploading and downloading the config, you could checkout the 2 repos in different directories, and then copy the config over

pavlovic-ivan replied:
Sounds good

ljubon replied:
FYI same was done for chart repo [here](https://github.com/ljubon/charts/blob/refactor-gh-pages/.github/workflows/publish-charts.yaml#L46) 👀

pavlovic-ivan replied:
@ljubon thanks a bunch! 🙇‍♂️

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,123 @@
+  PULUMI_CONFIG_PASSPHRASE: ${{ secrets.PULUMI_CONFIG_PASSPHRASE }}
+  GOOGLE_CREDENTIALS: ${{ secrets.GCE_SA_KEY }}
+  GOOGLE_REGION: ${{ secrets.GOOGLE_REGION }}
+  PULUMI_STACK_NAME: ${{ github.event.workflow_run.id }}
```
maybe we could prefix this with the repo name as well?

pavlovic-ivan replied:
Yes, we can. I'll add that as well

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,123 @@
+    if: ${{ github.event.action == 'completed' && needs.is-fork.outputs.upstream }}
+    environment: cuda-test-infra
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Prepare artifacts
+        run: |
+          mkdir -p artifacts
+          cp config.yaml artifacts
+
+      - name: Upload config
+        uses: actions/upload-artifact@v2
+        with:
+          name: pulumi-config
+          path: artifacts
+          retention-days: 1
+
+      - name: Checkout Pulumi Ephemeral Github Runner project
+        uses: actions/checkout@v2
+        with:
+          repository: pavlovic-ivan/ephemeral-github-runner-gcp
+          ref: main
+
+      - name: Download config
+        uses: actions/download-artifact@v2
+        with:
+          name: pulumi-config
+          path: artifacts
```
Same remark as above

pavlovic-ivan replied:
All right, thanks :)

jgiannuzzi commented on config.yaml:
```diff
@@ -0,0 +1,13 @@
+  ephemeral-github-runner:repo: ILGPU
+  ephemeral-github-runner:repoOwner: m4rs-mt
+  ephemeral-github-runner:runnersCount: 4
+  ephemeral-github-runner:serviceAccountEmail: pulumi@gr-oss-251320.iam.gserviceaccount.com
```
[optional] could this be made a secret?

pavlovic-ivan replied:
Yes, of course. I'll make it a secret

pavlovic-ivan commented on .github/workflows/ci.yaml:
```diff
@@ -6,12 +6,27 @@ on:
   DOTNET_NOLOGO: true
 
 jobs:
+  is-upstream:
+    runs-on: ubuntu-latest
+    environment: test-upstream-check
+    steps:
+      - name: Check if repo is upstream
+        id: check-if-upstream
+        run: |
+          fork=$(gh api repos/${{ github.repository }} | jq .fork)
+          upstream=false
+          [ $fork == "false" ] && upstream=true
+          echo "::set-output name=upstream::$upstream"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    outputs:
+      upstream: ${{ steps.check-if-upstream.outputs.upstream }}
```
I get what you mean. This is where my developer side kicked in. I don't like negation. Renaming this to `is-fork` means that double negation will be used as well down the line. To me, `is-fork` is equal to `is not upstream`. To check if it is upstream, i would have to do `not(is-fork)`, translating into `is not not upstream`. You know what i mean? 😄

jgiannuzzi replied:
I get your point, but here you are essentially implementing `is-upstream` as `not(is-fork)`, just so you can do `not(not(is-fork))` further. The 4 lines of shell would be just one if this was `is-fork`:

```yaml
run: echo "::set-output name=fork::$(gh api repos/${{ github.repository }} | jq .fork)"
```

pavlovic-ivan commented on .github/workflows/ci.yaml:
```diff
@@ -6,12 +6,27 @@ on:
     # Run daily at 00:00 so we get notified if CI is broken before a pull request
     # is submitted.
-    - cron:  '0 0 * * *'
+  - cron: '0 0 * * *'
```
Understood, thanks!

pavlovic-ivan commented on .github/workflows/ci.yaml:
```diff
@@ -72,6 +87,7 @@ jobs:
     # run the CI for each branch push to a fork, and for each pull request originating from a fork.
     if: github.event_name == 'schedule' || github.event_name == 'push' || github.event.pull_request.head.repo.id != github.event.pull_request.base.repo.id
     runs-on: ubuntu-latest
+    needs: is-upstream
```
Agreed

ljubon commented on .github/workflows/ci.yaml:
```diff
@@ -137,7 +146,7 @@ jobs:
             library: ILGPU.Algorithms
             framework: net471
-      fail-fast: true
+      fail-fast: false
```
is there any specific reason why this is set to `false`?

pavlovic-ivan replied:
This was an accident. I'll revert. Thanks for the sharp eye!

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,89 @@
+      - completed
+      - requested
+
+env:
+  PAT: ${{ secrets.PAT }}
+  GOOGLE_PROJECT: ${{ secrets.GOOGLE_PROJECT }}
+  PULUMI_CONFIG_PASSPHRASE: ${{ secrets.PULUMI_CONFIG_PASSPHRASE }}
+  GOOGLE_CREDENTIALS: ${{ secrets.GCE_SA_KEY }}
+  GOOGLE_REGION: ${{ secrets.GOOGLE_REGION }}
+  PULUMI_STACK_NAME: ilgpu_${{ github.event.workflow_run.id }}
+  GS_BUCKET: ${{ secrets.GS_BUCKET }}
+  GCE_SA_EMAIL: ${{ secrets.GCE_SA_EMAIL }}
```
I would avoid exposing all those secrets to the steps that don't need it. We could maybe move them to the `Execute pulumi` steps?

pavlovic-ivan replied:
Avreed

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,89 @@
+  GCE_SA_EMAIL: ${{ secrets.GCE_SA_EMAIL }}
+
+jobs:
+  is-fork:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check if repo is fork
+        id: is-fork
+        run: echo "::set-output name=fork::$(gh api repos/${{ github.repository }} | jq .fork)"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    outputs:
+      fork: ${{ steps.is-fork.outputs.fork }}
```
Could you please check whether this is necessary or whether the information is already somewhere under `${{ github }}`? Maybe it will still be necessary for `workflow_run` events triggered by `scheduled` workflow runs, in which case, there would be nothing to change!

jgiannuzzi replied:
See https://github.com/pavlovic-ivan/ILGPU/pull/4#discussion_r783767462 as well

pavlovic-ivan replied:
Will do, but i think this was already investigated, and the only reason why this is the implementation, is exactly as you've said, because of `workflow_run` events

pavlovic-ivan replied:
Ok, so i've checked the github context and here is the situations:

Upstream Repo - on manual push:
- main workflow -> `github.event.repository.fork = false`
- on workflow_run -> `github.event.repository.fork = false`

Upstream Repo - on scheduled run:
- main workflow -> `github.event.repository.fork = /` doesn't exist. github event object looks like this 
`github: {
  event: {
    schedule: */15 * * * *
  }
}`
- on workflow_run -> `github.event.repository.fork = false`

Forked repo - on manual push:
- main workflow -> `github.event.repository.fork = true`
- on workflow_run -> `github.event.repository.fork = true`

Forked repo - on scheduled run:
- main workflow -> `github.event.repository.fork = /` doesn't exist. github event object looks like this 
`github: {
  event: {
    schedule: */15 * * * *
  }
}`
- on workflow_run -> `github.event.repository.fork = true`

There is a bug apparently in the github.event context only for the main workflow. Event object is missing all kinds of data, it only contains the `schedule` fields set to what ever is set in the yaml file. I remember this issue when was adding cuda jobs to the scheduled runs. For the time being, i think this particular line of code will need to stay.

jgiannuzzi replied:
as `github.event.repository.fork` always has the right value for `workflow_run`, I would say we should get rid of this whole extra job

pavlovic-ivan replied:
Ah, yes. You are right. For this workflow the entire job is unnecessary. I didn't see that, even though i've written it down...

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,89 @@
+        uses: actions/checkout@v2
+        with:
+          repository: pavlovic-ivan/ephemeral-github-runner-gcp
+          ref: main
```
note for future improvement (_not_ for now): this should probably be a tag instead, so we use a stable version of the repo

pavlovic-ivan replied:
yes, correct, currently there are no tags in the mentioned repo

pavlovic-ivan replied:
@jgiannuzzi do you feel like we can resolve this conversation

jgiannuzzi replied:
sure!

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,89 @@
+        with:
+          node-version: '14'
+
+      - name: Execute Pulumi
+        run: |
+          cd pulumi
+          curl -fsSL https://get.pulumi.com | sh
+          npm install
+          pulumi login $GS_BUCKET
+          pulumi stack init -s $PULUMI_STACK_NAME
+          pulumi up --diff --config-file ../ilgpu/config.yaml -y
```
I would like this step to be split between installing pulumi, logging into pulumi, and then the actual deployment with pulumi

pavlovic-ivan replied:
sounds good!

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,89 @@
+          curl -fsSL https://get.pulumi.com | sh
+          npm install
+          pulumi login $GS_BUCKET
+          pulumi stack select $PULUMI_STACK_NAME
+          pulumi destroy --config-file artifacts/config.yaml -y
+          pulumi stack rm $PULUMI_STACK_NAME -y
```
These are the only different lines between this job and the `deploy-runners` job. I think we should make it all one job, and then have a different behaviour in the 'Deploy with Pulumi' step based on the action (`requested` or `completed`).

jgiannuzzi commented on config.yaml:
```diff
@@ -0,0 +1,12 @@
+  ephemeral-github-runner:ghRunnerName: orville
+  ephemeral-github-runner:machineImage: projects/gr-oss-251320/global/images/ubuntu-2004-focal-v20211202-ghrv22851
+  ephemeral-github-runner:machineType: a2-highgpu-1g
+  ephemeral-github-runner:repo: ILGPU
+  ephemeral-github-runner:repoOwner: m4rs-mt
```
nitpick: can those maybe be deduced at runtime?

pavlovic-ivan replied:
I am sure it can, however i would have to investigate. Either those vars can get overriden inline when executing Pulumi, or i will need to pass this via ENV vars

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,89 @@
+          npm install
+          pulumi login $GS_BUCKET
+          pulumi stack select $PULUMI_STACK_NAME
+          pulumi destroy --config-file artifacts/config.yaml -y
```
This is the wrong path the config file
```suggestion
          pulumi destroy --config-file ../ilgpu/config.yaml -y
```

pavlovic-ivan replied:
Thanks for pointing this out, i was sure i've updated the path to the file!

pavlovic-ivan commented:
@jgiannuzzi you will notice that the last step in the `manage-runners` job, that executes pulumi, is actually now running a `make` command. I've added Makefile to ephemeral runners pulumi repo, to act as a wrapper to all those nasty lines of pulumi commands. Hope you don't mind! Looking forward for the feedback. I think now all tasks are resolved 🙏

jgiannuzzi commented on .github/workflows/runners.yaml:
```diff
@@ -0,0 +1,76 @@
+          GCE_SA_EMAIL: ${{ secrets.GCE_SA_EMAIL }}
+          REPO: ${{ github.repository }}
+        working-directory: pulumi
+        run: |
+          if [[ ${{ github.event.action }} == 'requested' ]]; then
+            make up stack=$PULUMI_STACK_NAME config=../ilgpu/config.yaml auto-approve=-y
+          elif [[ ${{ github.event.action }} == 'completed' ]]; then
+            make down stack=$PULUMI_STACK_NAME config=../ilgpu/config.yaml auto-approve=-y
+          else
+            echo "Workflow failed. Unrecognized workflow event received: github.event.action=${{ github.event.action }}"
+            exit 1
+          fi
```
I think that this would be better expressed with a `case` which just sets the make target (`up` or `down`), followed by only one `make $target stack=...` command.
I also wonder whether we should actually treat the case of the unknown workflow event, as we explicitly filter them at the top of the file. If you really want to treat that case, you should at least make the error a proper workflow error (see the [docs](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message)).

pavlovic-ivan replied:
Gotcha. Ok, thanks, i will refactor it like that. For the error case, i will remove it entirely, as you've said, with workflow events it should be filtered out already

jgiannuzzi commented on config.yaml:
```diff
@@ -0,0 +1,10 @@
+config:
+  ephemeral-github-runner:bootDiskSizeInGB: 200
+  ephemeral-github-runner:bootDiskType: pd-balanced
+  ephemeral-github-runner:env: dev
```
I'm not sure what this does — is this required? should it be `dev`?

jgiannuzzi commented on config.yaml:
```diff
@@ -0,0 +1,10 @@
+  ephemeral-github-runner:bootDiskSizeInGB: 200
+  ephemeral-github-runner:bootDiskType: pd-balanced
+  ephemeral-github-runner:env: dev
+  ephemeral-github-runner:ghRunnerName: orville
```
I'm not sure what this does either — it this required? why is it `orville`?

pavlovic-ivan replied:
It is required as it will be in the instance name visible in GCP. I was trying to be a bit creative about the name, there's a sci fi show called Orville, so that's why :) If you think it's best to rename it to `ghrunner`, i can do that as well

jgiannuzzi replied:
understood — I think the new name you came up with is more descriptive 👍

jgiannuzzi commented on config.yaml:
```diff
@@ -0,0 +1,10 @@
+  ephemeral-github-runner:machineType: a2-highgpu-1g
+  ephemeral-github-runner:runnersCount: 6
+  ephemeral-github-runner:serviceAccountScopes: https://www.googleapis.com/auth/cloud-platform
+  ephemeral-github-runner:zone: europe-west4-a
```
as we set the region as a secret, should we have the zone? should we even specify it? should we not let GCP distribute the VM in the zone of its choice instead?

pavlovic-ivan replied:
I'll test that out, but if i remember correctly, it is necessary even though we have a region set

pavlovic-ivan replied:
@jgiannuzzi do you see any action points here or can we close this discussion?

jgiannuzzi replied:
could this be a GCP thing, e.g. is there maybe a special 'automatic' zone?

pavlovic-ivan replied:
@jgiannuzzi  Honestly, i am not sure. Maybe this can be a part of a future improvement? In the meantime, we investigate if this is possible. How do you feel about that?

pavlovic-ivan replied:
I'll add it as a secret. I feel that would be best :)

jgiannuzzi reviewed (commented):
Great improvements, well done @pavlovic-ivan!

pavlovic-ivan commented:
🎉 🎉 🎉 

Awesome! Let's make some people happy :)

pavlovic-ivan commented:
@jgiannuzzi @ljubon please also take a look at the top-most comment i've made, with instructions. I've updated it just now. I've removed the list of secrets that need to be set, as they will be provided offline securely. 🙏

ljubon reviewed (approved):
Awesome! 💯
